File: src/api/controllers/certificates.controller.ts
import { Request, Response, NextFunction } from "express";
import multer from "multer";
import {
  uploadCertificate,
  listCertificates,
  setCertificateActive,
} from "../../services/certificate.service";

// Multer: almacenar P12 en memoria (nunca en disco)
export const p12Upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 512 * 1024 }, // 512 KB máximo
  fileFilter: (_req, file, cb) => {
    if (
      file.mimetype === "application/x-pkcs12" ||
      file.originalname.endsWith(".p12") ||
      file.originalname.endsWith(".pfx")
    ) {
      cb(null, true);
    } else {
      cb(new Error("Solo se aceptan archivos .p12 o .pfx"));
    }
  },
}).single("certificate");

// ─── GET /certificates ─────────────────────────────────────────────
export async function listCertificatesHandler(
  _req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const certs = await listCertificates();
    res.json({ success: true, certificates: certs });
  } catch (err) {
    next(err);
  }
}

// ─── POST /certificates ────────────────────────────────────────────
// Body: multipart/form-data
//   certificate: <archivo .p12>
//   password: <contraseña del P12>
//   name: <nombre descriptivo>
export async function uploadCertificateHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    if (!req.file) {
      res.status(400).json({ success: false, error: "Se requiere el archivo P12" });
      return;
    }

    const { name, password } = req.body as { name?: string; password?: string };

    if (!password) {
      res.status(400).json({ success: false, error: "Se requiere la contraseña del P12" });
      return;
    }

    const cert = await uploadCertificate({
      name: name ?? req.file.originalname,
      p12Buffer: req.file.buffer,
      password,
    });

    // No retornar el P12 cifrado en la respuesta
    res.status(201).json({
      success: true,
      certificate: {
        id: cert.id,
        name: cert.name,
        ruc: cert.ruc,
        validFrom: cert.validFrom,
        validUntil: cert.validUntil,
        active: cert.active,
      },
    });
  } catch (err) {
    next(err);
  }
}

// ─── PATCH /certificates/:id/activate ────────────────────────────
export async function activateCertificateHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const cert = await setCertificateActive(req.params.id, true);
    res.json({ success: true, certificate: { id: cert.id, active: cert.active } });
  } catch (err) {
    next(err);
  }
}

// ─── PATCH /certificates/:id/deactivate ──────────────────────────
export async function deactivateCertificateHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const cert = await setCertificateActive(req.params.id, false);
    res.json({ success: true, certificate: { id: cert.id, active: cert.active } });
  } catch (err) {
    next(err);
  }
}

File: src/VenturacomSri.Api/Controllers/CertificatesController.cs
using Microsoft.AspNetCore.Mvc;
using VenturacomSri.Api.Services;

namespace VenturacomSri.Api.Controllers;

[ApiController]
[Route("api/v1/[controller]")]
public class CertificatesController : ControllerBase
{
    private readonly CertificateService _svc;

    public CertificatesController(CertificateService svc) => _svc = svc;

    // GET /api/v1/certificates
    [HttpGet]
    public async Task<IActionResult> List()
    {
        var certs = await _svc.ListAsync();
        return Ok(new { success = true, certificates = certs });
    }

    // POST /api/v1/certificates
    // multipart/form-data: certificate (.p12), password, name
    [HttpPost]
    [RequestSizeLimit(1_048_576)] // 1 MB
    public async Task<IActionResult> Upload(
        IFormFile certificate,
        [FromForm] string password,
        [FromForm] string? name)
    {
        if (certificate is null || certificate.Length == 0)
            return BadRequest(new { success = false, error = "Se requiere el archivo P12." });

        if (string.IsNullOrWhiteSpace(password))
            return BadRequest(new { success = false, error = "Se requiere la contraseña del P12." });

        var ext = Path.GetExtension(certificate.FileName).ToLowerInvariant();
        if (ext is not (".p12" or ".pfx"))
            return BadRequest(new { success = false, error = "Solo se aceptan archivos .p12 o .pfx." });

        using var ms = new MemoryStream();
        await certificate.CopyToAsync(ms);
        var p12Buffer = ms.ToArray();

        var cert = await _svc.UploadAsync(
            name ?? certificate.FileName,
            p12Buffer,
            password);

        return CreatedAtAction(nameof(List), null, new
        {
            success = true,
            certificate = new
            {
                cert.Id, cert.Name, cert.Ruc,
                cert.ValidFrom, cert.ValidUntil, cert.Active
            }
        });
    }

    // PATCH /api/v1/certificates/{id}/activate
    [HttpPatch("{id}/activate")]
    public async Task<IActionResult> Activate(string id)
    {
        var cert = await _svc.SetActiveAsync(id, true);
        return Ok(new { success = true, certificate = new { cert.Id, cert.Active } });
    }

    // PATCH /api/v1/certificates/{id}/deactivate
    [HttpPatch("{id}/deactivate")]
    public async Task<IActionResult> Deactivate(string id)
    {
        var cert = await _svc.SetActiveAsync(id, false);
        return Ok(new { success = true, certificate = new { cert.Id, cert.Active } });
    }
}
